File: server/fastapi/goban_server_fastapi/records/rest/update.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel
from sqlalchemy import select

from fastapi import Depends, HTTPException
from goban_server_fastapi.auth import User, jwt_user
from goban_server_fastapi.db import DbClient, dictify
from goban_server_fastapi.records.models import Record
from goban_server_fastapi.rest import app


class UpdateRequest(BaseModel):
    name: str
    black_player: str
    white_player: str
    comment: str
    handicap: int
    komi: float
    ruleset: Literal["AGA", "JPN", "CHN"]
    winner: Literal["U", "B", "W"]


class UpdateResponse(BaseModel):
    id: int
    owner: int
    board_size: Literal[9, 13, 19]
    created: datetime
    name: str
    black_player: str
    white_player: str
    comment: str
    handicap: int
    komi: float
    ruleset: Literal["AGA", "JPN", "CHN"]
    winner: Literal["U", "B", "W"]


@app.put(
    "/api/records/{record_id}/",
    status_code=200,
    response_model=UpdateResponse,
)
def update_record(
    record_id: int,
    record_model: UpdateRequest,
    db: DbClient = Depends(),
    current_user: User = Depends(jwt_user),
):
    with db.session() as session:
        record: Record = session.scalar(
            select(Record)
            .where(Record.id == record_id)
            .where(Record.owner_id == current_user.id)
        )
        if record is None:
            raise HTTPException(status_code=404)
        for key, value in record_model.dict().items():
            setattr(record, key, value)
        session.commit()
        return {
            **dictify(record),
            "owner": record.owner_id,
        }

File: server/deno/src/record/update.ts
import { register, validator } from "/router.ts";
import { sql } from "/db.ts";
import { json_schema as J } from "/deps.ts";
import { getUserId } from "/auth/util.ts";

const UpdateRequest = J.struct({
  name: J.string(),
  black_player: J.string(),
  white_player: J.string(),
  comment: J.string(),
  handicap: J.number(),
  komi: J.number(),
  ruleset: J.string(),
  winner: J.string(),
});
const body = validator(UpdateRequest);

register("PUT", "/api/records/{id}/", async (request, { id: recordId }) => {
  const userId = await getUserId(request);
  const {
    name,
    black_player,
    white_player,
    comment,
    handicap,
    komi,
    ruleset,
    winner,
  } = await body(request);
  const rows = await sql`UPDATE record_record SET
  name=${name}, black_player=${black_player}, white_player=${white_player}, comment=${comment}, handicap=${handicap}, komi=${komi}, ruleset=${ruleset}, winner=${winner}
  WHERE
  owner_id=${userId} AND id=${recordId}
  RETURNING board_size, created`;
  const board_size = Number.parseInt(rows[0].board_size, 10);
  const created = rows[0].created;
  return new Response(
    JSON.stringify({
      id: Number.parseInt(recordId, 10),
      owner: userId,
      board_size,
      created,
      name,
      black_player,
      white_player,
      comment,
      handicap,
      komi,
      ruleset,
      winner,
    }),
  );
});
